src/db/api/abstract.ts

```diff
@@ -0,0 +1,24 @@
+import {IDatabasePersistance} from "../persistance";
+import {Service} from "../../node";
+
+export interface DatabaseApiOptions {
+  persistance: IDatabasePersistance;
+}
+
+export abstract class DatabaseApi implements Service{
+
+  protected db: IDatabasePersistance;
```

Comment:
Yeah, add cache service here, and use it where it makes sense I guess.

Reply:
How would the cache be implemented?

Reply:
I would approached this with something like this:
```
protected cache: ICacheService;

getState() {
   const key = encodeKey(Bucket.chainInfo, Key.state);
   return await this.cache.getOrUpdate(key, (key) => this.db.getState(), expiresAt = null);
}
```

Basically cache service would contain map object and with this method it will check if it contains object with that key (and is not expired) it would return that element otherwise it would execute callback, update key with callback result and return result from function.

I mean there is plenty of options, I liked this approach from when I was doing laravel apps (https://laravel.com/docs/5.8/cache#retrieving-items-from-the-cache) :smile:

Reply:
There would also be method to update state forcefully (not only when it expires). For example, you would invoke `this.cache.put(key, state)` on `setState` method.